### PR TITLE
Deprecation of AutoML services: Add deprecation warnings and raise exceptions for already deprecated ones

### PR DESCRIPTION
**Pull request by molcay**
> Before merging this PR we need to merge #38633 

This PR is for raising exception for the deprecated services of [AutoML](https://cloud.google.com/automl/docs) ([Tables](https://cloud.google.com/automl-tables/docs/deprecations), [Vision](https://cloud.google.com/vision/automl/docs/deprecations), [Video Intelligence](https://cloud.google.com/video-intelligence/automl/docs/deprecations), [Natural Language](https://cloud.google.com/natural-language/automl/docs/deprecations)) which are EoL or will be EoL soon.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/google/cloud/hooks/vertex_ai/prediction_service.py
Comment (eladkal) [inline airflow/providers/google/cloud/hooks/vertex_ai/prediction_service.py:34]: You will need to add tests for this hook

**Comment by potiuk** [comment]
I think you need to rebase the PR. It's 52 commits behind and this is likely why it fails

File: tests/providers/google/cloud/operators/test_automl.py
Comment (Taragolis) [inline tests/providers/google/cloud/operators/test_automl.py]: warnings intends to be checked by the [`pytest.warns`](https://docs.pytest.org/en/stable/how-to/capture-warnings.html#warns)

Comment (molcay) [reply tests/providers/google/cloud/operators/test_automl.py]: @Taragolis I was not aware of this one. Thank you for the warning. I am on it

Comment (molcay) [reply tests/providers/google/cloud/operators/test_automl.py]: @Taragolis Hi again, I checked the docs and tried to use warnings (`pytest.warns`) but the thing is that we are **raising** the `AirflowProviderDeprecationWarning` in case of specific usage of the operators. I know it is a little bit different than general behavior but all these services ([AutoML](https://cloud.google.com/automl/docs) ([Tables](https://cloud.google.com/automl-tables/docs/deprecations), [Vision](https://cloud.google.com/vision/automl/docs/deprecations), [Video Intelligence](https://cloud.google.com/video-intelligence/automl/docs/deprecations), [Natural Language](https://cloud.google.com/natural-language/automl/docs/deprecations))) already are EoL but same operators can be use for AutoML Translation which is **NOT** deprecated.

File: airflow/providers/google/cloud/hooks/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/hooks/automl.py:653]: Could you please add a docstring and tests for this method?

File: airflow/providers/google/cloud/hooks/vertex_ai/prediction_service.py
Comment (shahar1) [inline airflow/providers/google/cloud/hooks/vertex_ai/prediction_service.py]: Same as my previous comment :)

File: airflow/providers/google/cloud/hooks/vertex_ai/prediction_service.py
Comment (shahar1) [inline airflow/providers/google/cloud/hooks/vertex_ai/prediction_service.py]: Could you please recheck the `params` argument?
According to the API, if I'm not wrong, it should be named `parameters`.

File: airflow/providers/google/cloud/operators/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/operators/automl.py]: Following @Taragolis 's comment, you should either `raise AirflowException("...")` or `warnings.warn("...", AirflowProviderDeprecationWarning)` - as these are the conventions for applying exceptions and warnings, respectively.

File: airflow/providers/google/cloud/operators/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/operators/automl.py]:
```suggestion
        f"Please use {alternative_class_name_str} from Vertex AI."
```

File: airflow/providers/google/cloud/operators/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/operators/automl.py]: Please update similar warnings as well, as it has already been removed by GCP:

```suggestion
        f"{deprecated_class_name} for text, image, and video prediction has been "
        f"deprecated and no longer available. All the functionality of "
```

File: airflow/providers/google/cloud/operators/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/operators/automl.py:237]: Please add a docstring for the new argument

File: airflow/providers/google/cloud/operators/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/operators/automl.py:276]: Please add tests for this `if`'s causes

File: airflow/providers/google/cloud/operators/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/operators/automl.py]: Could you please add a test for this exception?

File: airflow/providers/google/cloud/operators/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/operators/automl.py]: 1. `model_id` is a template field; related value checks shouldn't be here - but in the beginning of the `execute()` method. Tou could relocate this `if` there, while referring to `model_Id` as `self.model_id` and `endpoint_id` as `self.endpoint_id`.
2. Please add a test for this `if` after relocating it.

File: airflow/providers/google/cloud/operators/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/operators/automl.py]: Why don't you trust the return type hints given in the methods' signature?

```suggestion
    def _check_model_type(self):
        hook = self.hook
        model = hook.get_model(
```

File: airflow/providers/google/cloud/operators/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/operators/automl.py]:
```suggestion
        hook = self.hook
```

File: airflow/providers/google/cloud/operators/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/operators/automl.py:327]: Why is `self.model_id` now required?

File: tests/providers/google/cloud/operators/test_automl.py
Comment (shahar1) [inline tests/providers/google/cloud/operators/test_automl.py]: Instead of having these `#noqa: E402`, you could try adding a line in `pyproject.toml` (around L356)

File: tests/providers/google/cloud/operators/test_automl.py
Comment (shahar1) [inline tests/providers/google/cloud/operators/test_automl.py]: It would be nicer to use the `match` keyword instead:
https://docs.pytest.org/en/7.1.x/how-to/assert.html

You are welcome to apply it to similar assertions :)

File: airflow/providers/google/cloud/operators/automl.py
Comment (shahar1) [inline airflow/providers/google/cloud/operators/automl.py]: Please add a test for this `if` (and similar `if`s that raise an exception)

**Review by shahar1 — CHANGES_REQUESTED** [review]
Good start! 
There's some work to do to finish it up (mostly adding tests and docstrings), and I hope that my comments will make sense - please ask if you have any questions.
I'd be happy if you could test the operator against an actual GCP project with enabled AutoML API to see if all functionalities work as expected :)

File: tests/providers/google/cloud/operators/test_automl.py
Comment (Taragolis) [reply tests/providers/google/cloud/operators/test_automl.py]: We should not raise warning as regular exception and usual it is semantically incorrect, because this is user decision whether or not raise specific warning as exception by configure `PYTHONWARNINGS` (https://docs.python.org/3/library/warnings.html)

Comment (Taragolis) [reply tests/providers/google/cloud/operators/test_automl.py]: Right now you test works because from recent changes in all new tests prohibit by specific warnings, see: https://github.com/apache/airflow/blob/main/contributing-docs/testing/unit_tests.rst#handling-warnings

But in user runtime it will not raise any error, only the warning.

File: tests/providers/google/cloud/operators/test_automl.py
Comment (molcay) [reply tests/providers/google/cloud/operators/test_automl.py]: I was not aware of this feature. Thank you for the information! I am changing accordingly

File: airflow/providers/google/cloud/hooks/vertex_ai/prediction_service.py
Comment (molcay) [reply airflow/providers/google/cloud/hooks/vertex_ai/prediction_service.py]: Nice catch! I changed it as `parameters`

**Comment by potiuk** [comment]
@molcay -> can you please rebase/resolve conflict and respond/mark as resolved all conversations that you think were resolved?

**Comment by potiuk** [comment]
cc: @VladaZakharova - there were some doubts  for Google team about those deprecations - does this one look ok for you?

**Comment by VladaZakharova** [comment]
> cc: @VladaZakharova - there were some doubts for Google team about those deprecations - does this one look ok for you?

Hi! Thanks, yes, changes look good, LGTM

**Review by VladaZakharova — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
@molcay -> can you please review and address/mark as resolved if they are - the comments made by @shahar1 ?

**Comment by molcay** [comment]
@potiuk,
Ohh I see, there are some hidden conversations. I totally missed them. Sorry for the inconvenience. I will try to address and resolve them as soon as possible.

File: airflow/providers/google/cloud/operators/automl.py
Comment (molcay) [reply airflow/providers/google/cloud/operators/automl.py:327]: Since we introduce the `endpoint_id`, `model_id` is optional right now. This check is to ensure the `model_id` is given. Because `AutoMLModelPredictLink` can be used only for `model_id` case

**Review by shahar1 — APPROVED** [review]
Very good progress! :)
You resolved all of my comments, so you get a sincere approval from me.
One last step before merging - try to rebase and ensure that `tests/providers/google/cloud/operators/test_automl.py` passes in the Non-DB tests, as currently they fail on the CI.
Make sure that objects are mocked properly in tests that you added or modified.

**Comment by molcay** [comment]
I will work on this test failure and send an update. Thank you for the review @shahar1 
Also, thank you for pinging and guiding along the way @potiuk

**Review by eladkal — APPROVED** [review]

**Comment by VladaZakharova** [comment]
Hey @shahar1 !
Would be nice if you can check the changes here again

**Review by shahar1 — APPROVED** [review]
Re-reviewed, we're good to go :)

**Pull request merged**
